### PR TITLE
Support Dictionary Based Plan For DISTINCT

### PR DESCRIPTION
**Pull request by atris**
## Description
This PR introduces the ability to execute DISTINCT using a dictionary based plan when the DISTINCT query has a single column, no filters and the column is dictionary encoded.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by atris** [comment]
@mayankshriv @Jackie-Jiang Please review

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 1]
```diff
@@ -264,4 +267,31 @@ static boolean isFitForDictionaryBasedPlan(QueryContext queryContext, IndexSegme
...
+    DistinctAggregationFunction distinctAggregationFunction = (DistinctAggregationFunction) aggregationFunctions[0];
+    List<ExpressionContext> expressions = distinctAggregationFunction.getInputExpressions();
+
+    if (expressions.size() == 1 && queryContext.getFilter() == null) {
+      ExpressionContext expression = expressions.get(0);
+
+      if (expression.getType() == ExpressionContext.Type.IDENTIFIER) {
+        String column = expression.getIdentifier();
+        Dictionary dictionary = indexSegment.getDataSource(column).getDictionary();
+        if (dictionary != null && dictionary.isSorted()) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:289]: We should still be able to use dictionary based operator when dictionary is not sorted: create a `DistinctTable` and add all dictionary values into the table

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        for (int dictId = 0; dictId < dictionarySize; dictId++) {
+            if (!_hasOrderBy) {
+                _dictIdSet.add(dictId);
+
+                if (_dictIdSet.size() >= limit) {
+                    break;
+                }
+            } else {
+                // We already ascertained that the dictionary is sorted, hence we can use IDs instead of actual values
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:102]: Since we already know the dictionary is sorted, simply pick the first (asc) or last (desc) `limit` dictIds and put them into the `DistinctTable`, no need to scan the dictionary.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        int limit = _distinctAggregationFunction.getLimit();
+        String column = _distinctAggregationFunction.getInputExpressions().get(0).getIdentifier();
+
+        assert _distinctAggregationFunction.getType() == AggregationFunctionType.DISTINCT;
+
+        Dictionary dictionary = _dictionaryMap.get(column);
+        int dictionarySize = dictionary.length();
+
+        for (int dictId = 0; dictId < dictionarySize; dictId++) {
+            if (!_hasOrderBy) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:95]: Without ordering, pick the first `limit` dictIds and put them into the `DistinctTable`

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 4]
```diff
@@ -264,4 +267,31 @@ static boolean isFitForDictionaryBasedPlan(QueryContext queryContext, IndexSegme
     }
     return true;
   }
+
+  /**
+   * Returns dictionary based distinct plan node iff supported, else distinct plan node
+   */
+  private PlanNode getDistinctExecutionNode(IndexSegment indexSegment, QueryContext queryContext) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:274]: (nit) `getDistinctPlanNode`

File: pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    public DictionaryBasedDistinctPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
+        _indexSegment = indexSegment;
+        AggregationFunction[] aggregationFunctions = queryContext.getAggregationFunctions();
+
+        assert aggregationFunctions != null && aggregationFunctions.length == 1
+                && aggregationFunctions[0] instanceof DistinctAggregationFunction;
+
+        _distinctAggregationFunction = (DistinctAggregationFunction) aggregationFunctions[0];
+
+        _dictionaryMap = new HashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java:54]: We don't really need a map. There should be only one dictionary. You can directly pass in the dictionary when constructing the plan node

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java [hunk 6]
```diff
@@ -245,19 +245,60 @@ public void testSingleColumnDistinctOnlyInnerSegment()
       // String columns
       //@formatter:off
       List<String> queries = Arrays
-          .asList("SELECT DISTINCT(stringColumn) FROM testTable", "SELECT DISTINCT(rawStringColumn) FROM testTable");
+          .asList("SELECT DISTINCT(stringColumn) FROM testTable ORDER BY stringColumn");
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:248]: This test is already covered in `testSingleColumnDistinctOrderByInnerSegment()`. We don't need to add extra correctness test for the change, but you may want to add test to ensure the dictionary-based operator is used

**Review by Jackie-Jiang — COMMENTED** [review]
The code structure is good. We can utilize the property of the dictionary (sorted, no duplicate values) to further optimize the operator

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java (hunk 1 above)
Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:289]: How would we apply limits on it?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java (hunk 3 above)
Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:95]: Isnt that what we are doing here?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java (hunk 2 above)
Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:102]: That would not work,  since dictionary does not allow us to get dictIds from it directly. We would have to get it from TransformBlock, which does not guarantee order of returned dictIds.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:102]: I don't follow. dictId is simply the index within dictionary. E.g. for ascending order of limit 10, just read dictId 0-9 from the dictionary and construct the DistinctTable. No need to scan the whole dictionary

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:95]: No, currently we scan the whole dictionary. Instead, we only need to scan `limit` number of entries in the dictionary

**Comment by atris** [comment]
@Jackie-Jiang Updated, please see and let me know your thoughts

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.apache.pinot.spi.data.FieldSpec;
+
+/**
+ * Operator which executes DISTINCT operation based on dictionary
+ */
+public class DictionaryBasedDistinctOperator extends DistinctOperator {
+    private static final String OPERATOR_NAME = "DictionaryBasedDistinctOperator";
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:53]: The format is still incorrect

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java [hunk 8]
```diff
@@ -245,12 +245,14 @@ public void testSingleColumnDistinctOnlyInnerSegment()
       // String columns
       //@formatter:off
       List<String> queries = Arrays
-          .asList("SELECT DISTINCT(stringColumn) FROM testTable", "SELECT DISTINCT(rawStringColumn) FROM testTable");
+          .asList("SELECT DISTINCT(stringColumn) FROM testTable");
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:248]: Revert the change in this file

File: pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        AggregationFunction[] aggregationFunctions = queryContext.getAggregationFunctions();
+
+        assert aggregationFunctions != null && aggregationFunctions.length == 1
+                && aggregationFunctions[0] instanceof DistinctAggregationFunction;
+
+        _distinctAggregationFunction = (DistinctAggregationFunction) aggregationFunctions[0];
+
+        _dictionary = dictionary;
+
+        _transformPlanNode =
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java:54]: We don't need to construct the transform plan node. Dictionary itself is enough to solve the query

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    @Override
+    protected IntermediateResultsBlock getNextBlock() {
+        int limit = _distinctAggregationFunction.getLimit();
+        ExpressionContext expression = _distinctAggregationFunction.getInputExpressions().get(0);
+
+        assert _distinctAggregationFunction.getType() == AggregationFunctionType.DISTINCT;
+
+        TransformBlock transformBlock = _transformOperator.nextBlock();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:96]: We don't need to read dictIds from the transform block.

- If there is no order-by, read the first `min(limit, dictionarySize)` values from the dictionary
- If there is order-by
  - If dictionary is sorted, read the first if asc, last if desc `min(limit, dictionarySize)` values from the dictionary
  - If dictionary is not sorted, dump all values in the dictionary into the `DistinctTable`

Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:96]: Yes, I was not aware that dictionary (when sorted) has IDs strictly in the ascending order of numbers (i.e. starting dictID will always be 0, next 1 ...). Thanks for highlighting that. Fixed per your suggestions.

For the no order by case, we dont really have an order defined in which we read values from the underlying storage (so the order of values read from segment vs dictionary can be different). Hence, need to change DistinctQueriesTest to reflect the same.

File: pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java (hunk 9 above)
Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java:54]: We need the transform plan node to get the data type during result construction

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java (hunk 8 above)
Comment (atris) [reply pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:248]: We need the changes since the order of results in absence of ORDER BY is not consistent between segment reads and dictionary reads.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:47]: I applied the style sheet and it seems to be passing -- what is the formatting change needed here, please?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java (hunk 2 above)
Comment (atris) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java]: Yeah, as discussed, I was not aware that dictIDs are strictly in the order of 0 - dictionary limit. Changed now, thanks

File: pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java (hunk 9 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java:54]: Data type is available in by `Dictionary.getValueType()`. We want to short-circuit all the lower level operators

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        List<ExpressionContext> expressions = _distinctAggregationFunction.getInputExpressions();
+        ExpressionContext expression = expressions.get(0);
+        FieldSpec.DataType dataType = _transformOperator.getResultMetadata(expression).getDataType();
+
+        DataSchema dataSchema = new DataSchema(new String[]{expression.toString()},
+                new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.fromDataTypeSV(dataType)});
+        List<Record> records;
+
+        int limit = _distinctAggregationFunction.getLimit();
+        int actualLimit = Math.min(limit, _dictionary.length());
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:98]: Let's cache the value of `_dictionary.length()` into a local variable

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                    for (int i = 0; i < actualLimit; i++) {
+                        records.add(new Record(new Object[]{_dictionary.getInternal(i)}));
+                    }
+                } else {
+                    for (int i = _dictionary.length() - 1; i >= (_dictionary.length() - actualLimit); i--) {
+                        records.add(new Record(new Object[]{_dictionary.getInternal(i)}));
+                    }
+                }
+            } else {
+                records = new ArrayList<>(_dictionary.length());
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:122]: We should insert records into the `DistinctTable`. `DistinctTable` will only keep the top records.
```suggestion
  DistinctTable distinctTable = new DistinctTable(dataSchema, _distinctAggregationFunction.getOrderByExpressions(), limit);
  for (int i = 0; i < dictionarySize; i++) {
    distinctTable.addWithOrderBy(new Record(new Object[]{_dictionary.getInternal(i)}));
  }
```

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java (hunk 8 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:248]: I see. This is caused by sorting in alphabetical order. Can you please add some comments explaining the reason?

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java [hunk 13]
```diff
@@ -245,7 +245,34 @@ public void testSingleColumnDistinctOnlyInnerSegment()
       // String columns
       //@formatter:off
       List<String> queries = Arrays
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:247]: Since there is only one query, let's use a single value instead of a list. We can also remove the comment to turn off/on the formatter

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java [hunk 14]
```diff
@@ -245,7 +245,34 @@ public void testSingleColumnDistinctOnlyInnerSegment()
...
-          .asList("SELECT DISTINCT(stringColumn) FROM testTable", "SELECT DISTINCT(rawStringColumn) FROM testTable");
+          .asList("SELECT DISTINCT(stringColumn) FROM testTable");
+      //@formatter:on
+      Set<Integer> expectedValues =
+          new HashSet<>(Arrays.asList(0, 16, 17, 1, 10, 11, 12, 13, 14, 15));
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:251]: (nit) for clarity (alphabetical order)
```suggestion
          new HashSet<>(Arrays.asList(0, 1, 10, 11, 12, 13, 14, 15, 16, 17));
```

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java [hunk 15]
```diff
@@ -245,7 +245,34 @@ public void testSingleColumnDistinctOnlyInnerSegment()
...
+            actualValues.add(Integer.parseInt((String) values[0]));
+          }
+          assertEquals(actualValues, expectedValues);
+        }
+      }
+    }
+    {
+      // String columns
+      //@formatter:off
+      List<String> queries = Arrays
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:274]: Same here, use single value

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.apache.pinot.spi.data.FieldSpec;
+
+
+/**
+ * Operator which executes DISTINCT operation based on dictionary
+ */
+public class DictionaryBasedDistinctOperator extends DistinctOperator {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:43]: We should extends `BaseOperator<IntermediateResultsBlock>` instead of `DistinctOperator`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+
+  @Override
+  public String getOperatorName() {
+    return OPERATOR_NAME;
+  }
+
+  @Override
+  public ExecutionStatistics getExecutionStatistics() {
+    // NOTE: Set numDocsScanned to numTotalDocs for backward compatibility.
+    return new ExecutionStatistics(_numTotalDocs, 0, 0, _numTotalDocs);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:144]: Let's count the actual values read from the dictionary, and put it as `numDocsScanned` and `numEntriesScannedPostFilter`

File: pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    AggregationFunction[] aggregationFunctions = queryContext.getAggregationFunctions();
+
+    assert aggregationFunctions != null && aggregationFunctions.length == 1
+        && aggregationFunctions[0] instanceof DistinctAggregationFunction;
+
+    _distinctAggregationFunction = (DistinctAggregationFunction) aggregationFunctions[0];
+
+    _dictionary = dictionary;
+
+    _transformPlanNode =
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/DictionaryBasedDistinctPlanNode.java:54]: No need to create the transform plan

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+          for (int i = 0; i < actualLimit; i++) {
+            distinctTable.addWithOrderBy(new Record(new Object[]{_dictionary.getInternal(i)}));
+          }
+        } else {
+          for (int i = _dictLength - 1; i >= (_dictLength - actualLimit); i--) {
+            distinctTable.addWithOrderBy(new Record(new Object[]{_dictionary.getInternal(i)}));
+          }
+        }
+      } else {
+        for (int i = 0; i < _dictLength; i++) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:125]: You only need to create main distinct table when dictionary is not sorted. When dictionary is sorted, creating a wrapper distinct table is good enough

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 20]
```diff
@@ -264,4 +267,31 @@ static boolean isFitForDictionaryBasedPlan(QueryContext queryContext, IndexSegme
...
+    DistinctAggregationFunction distinctAggregationFunction = (DistinctAggregationFunction) aggregationFunctions[0];
+    List<ExpressionContext> expressions = distinctAggregationFunction.getInputExpressions();
+
+    if (expressions.size() == 1 && queryContext.getFilter() == null) {
+      ExpressionContext expression = expressions.get(0);
+
+      if (expression.getType() == ExpressionContext.Type.IDENTIFIER) {
+        String column = expression.getIdentifier();
+        Dictionary dictionary = indexSegment.getDataSource(column).getDictionary();
+        if (dictionary != null) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:289]: Per the test failure, we need to read the data type from the data source by calling `DataSource.getDataSourceMetadata().getDataType()`. The value type from dictionary is the internal stored type

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 21]
```diff
@@ -264,4 +267,31 @@ static boolean isFitForDictionaryBasedPlan(QueryContext queryContext, IndexSegme
...
+          return new DictionaryBasedDistinctPlanNode(indexSegment, queryContext, dictionary);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:290]: (nit) get the data type here and pass it in instead of passing in the segment

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+/**
+ * Operator which executes DISTINCT operation based on dictionary
+ */
+public class DictionaryBasedDistinctOperator extends BaseOperator<IntermediateResultsBlock> {
+  private static final String OPERATOR_NAME = "DictionaryBasedDistinctOperator";
+
+  private final DistinctAggregationFunction _distinctAggregationFunction;
+  private final Dictionary _dictionary;
+  private final int _numTotalDocs;
+  private IndexSegment _indexSegment;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:49]: This can be final

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+
+  private boolean _hasOrderBy;
+  private boolean _isAscending;
+
+  private int _dictLength;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java:54]: This can be final, or move to local variable (preferable)

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments
